revoke/revoke.go

```diff
@@ -123,69 +213,141 @@ func getIssuer(cert *x509.Certificate) *x509.Certificate {
+	if force || !r.isInMemoryCRLValid(newLocalCRL) {
+		if _, err := os.Stat(newLocalCRL); err != nil {
+			return fmt.Errorf("failed to read local CRL path: %v", err)
+		}
+
+		tmp, err := ioutil.ReadFile(newLocalCRL)
+		if err != nil {
+			return fmt.Errorf("failed to read local CRL path: %v", err)
+		}
+		crl, err := x509.ParseCRL(tmp)
+		if err != nil {
+			return fmt.Errorf("failed to parse local CRL file: %v", err)
 		}
+
+		r.lock.Lock()
+		r.crlSet[newLocalCRL] = crl
+		if r.localCRL != newLocalCRL {
+			if _, ok := r.crlSet[r.localCRL]; ok {
+				delete(r.crlSet, r.localCRL)
+			}
+			r.localCRL = newLocalCRL
+		}
+		r.lock.Unlock()
 	}
 
+	return nil
+}
+
+// FetchRemoteCRL fetches remote CRL into internal map,
+// force overwrites previously read CRL
+func (r *Revoke) FetchRemoteCRL(url string, cert *x509.Certificate, force bool) error {
+	shouldFetchCRL := !r.isInMemoryCRLValid(url)
+
 	issuer := getIssuer(cert)
 
-	if shouldFetchCRL {
+	if force || shouldFetchCRL {
 		var err error
-		crl, err = fetchCRL(url)
+		crl, err := fetchCRL(url)
 		if err != nil {
-			log.Warningf("failed to fetch CRL: %v", err)
-			return false, false
+			return fmt.Errorf("failed to fetch CRL: %v", err)
 		}
 
 		// check CRL signature
 		if issuer != nil {
 			err = issuer.CheckCRLSignature(crl)
 			if err != nil {
-				log.Warningf("failed to verify CRL: %v", err)
-				return false, false
+				return fmt.Errorf("failed to verify CRL: %v", err)
 			}
 		}
 
-		CRLSet[url] = crl
+		r.lock.Lock()
+		r.crlSet[url] = crl
+		r.lock.Unlock()
+	}
+
+	return nil
+}
+
+// check a cert against a specific CRL. Returns the same bool pair
+// as revCheck. If remote is false - will assume that url is a file.
+func (r *Revoke) certIsRevokedCRL(cert *x509.Certificate, url string, remote bool) (revoked, ok bool) {
+	var err error
+	if remote {
+		err = r.FetchRemoteCRL(url, cert, false)
+	} else {
+		err = r.fetchLocalCRL(url, false)
+	}
+
+	if err != nil {
+		log.Warningf("%v", err)
+		return false, false
 	}
 
-	for _, revoked := range crl.TBSCertList.RevokedCertificates {
+	defer r.lock.Unlock()
+	r.lock.Lock()
+	for _, revoked := range r.crlSet[url].TBSCertList.RevokedCertificates {
 		if cert.SerialNumber.Cmp(revoked.SerialNumber) == 0 {
-			log.Info("Serial number match: intermediate is revoked.")
 			return true, true
 		}
 	}
 
 	return false, true
 }
 
+// verifyCertTime verifies whether certificate time frames are valid
+func verifyCertTime(cert *x509.Certificate) bool {
+	if !time.Now().Before(cert.NotAfter) {
+		log.Infof("Certificate expired %s", cert.NotAfter)
+		return false
+	} else if !time.Now().After(cert.NotBefore) {
+		log.Infof("Certificate isn't valid until %s", cert.NotBefore)
+		return false
+	}
+
+	return true
+}
+
 // VerifyCertificate ensures that the certificate passed in hasn't
 // expired and checks the CRL for the server.
+// Comparing to the next public method, this function uses
```

Comment:
I thing godoc has its own way of ordering and formatting documents, so this comment is unnecessary and confusing, should be removed.